RFC: typed array conveniences

dherman opened the pull request:
This RFC defines a number of missing convenience APIs for typed arrays: constructors and metadata queries.

Implementation: neon-bindings/neon#909

[Rendered](https://github.com/neon-bindings/rfcs/blob/b400353cbbe41693094d5cda62b909e63482b919/text/0000-typed-array-conveniences.md)

dherman committed:
New RFC: typed array conveniences

kjvalencik commented on text/0000-typed-array-conveniences.md:
````diff
@@ -0,0 +1,312 @@
+The `JsArrayBuffer` type gets a new method for constructing a `Region` struct, which describes a region of a buffer (see below).
+
+```rust
+impl Handle<'cx, JsArrayBuffer> {
+    pub fn region<T: Binary>(self, byte_offset: usize, len: usize) -> Region<'cx, T>;
````
Should this be `to_region` for symmetry with the other methods? Also, I think `&self` is better unless there's a benefit of `self`.

dherman replied:
What if went the other direction, and had them both called `.region()`? In both cases it's not really a conversion, it's a kind of accessor.

kjvalencik replied:
I like `region`. Seeing it as an accessor (which should just be the name of the field, per API guidelines), this works perfectly.

kjvalencik commented on text/0000-typed-array-conveniences.md:
```diff
@@ -0,0 +1,312 @@
+
+### Helper types
+
+#### `neon::types::binary::Binary`
```
I think it would be better to put `Binary` and `Region` in the existing `neon::types::buffer` module to keep the tree slightly smaller and to co-locate array buffer related traits/types.

https://docs.rs/neon/latest/neon/types/buffer/index.html

dherman replied:
Whoops, that was just a brain-o when I typed up the RFC (the implementation PR already agrees with your suggestion). I'll fix the doc.

kjvalencik commented on text/0000-typed-array-conveniences.md:
```diff
@@ -0,0 +1,312 @@
+    // Constructs a new typed array for this region.
+    pub fn to_typed_array<'c, C>(self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
+    where C: Context<'c>;
+}
```
I think a nice addition would be to add a `to_region` on `Region` for nice chaining sub-regions and converting types.

```rust
impl<'cx, T: Binary> Region<'cx, T> {
    // Constructs a new typed array for this region.
    pub fn to_typed_array<'c, C>(&self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
    where
        C: Context<'c>;

    // Constructs a new sub-region
    pub fn to_region<'c, U, C>(&self, byte_offset: usize, len: usize) -> Region<'cx, U>
    where
        C: Context<'c>,
        U: Binary;
}
```

```rust
let numbers = cx.argument::<JsUint8Array>(0)?
    .to_region(&mut cx)
    .to_region<JsUint32Array, _>(&mut cx, 2, 4)
    .to_typed_array(&mut cx)?;
```

kjvalencik commented on text/0000-typed-array-conveniences.md:
```diff
@@ -0,0 +1,312 @@
+
+impl<'cx, T: Binary> Region<'cx, T> {
+    // Constructs a new typed array for this region.
+    pub fn to_typed_array<'c, C>(self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
```
```suggestion
    pub fn to_typed_array<'c, C>(&self, cx: &mut C) -> JsResult<'c, JsTypedArray<T>>
```

Methods prefix `to_` should take `&self`. Methods that take `self` should be named `into`. Since we're not really consuming anything, I think `&self` and keeping the method makes more sense.

dherman replied:
Makes sense. It's a bit of a theoretical distinction here since `Region` is `Copy`, but that might still be confusing from a documentation perspective, so I like your suggestion.

kjvalencik commented on text/0000-typed-array-conveniences.md:
```diff
@@ -0,0 +1,312 @@
+
+## Naming inconsistency: `len()` vs `byte_length()`
+
+This RFC goes with `len()` to match a strong Rust convention, and `byte_length()` to match the typed array naming convention (using snake case, of course). These does lead to an inconsistency between the two. A couple alternatives would be `byte_len()` or `size()`. But then `byte_len()` ends up feeling a bit inconsistent with `byte_offset()`. We could then choose `byte_off()` but that's a bit opa ...
```
I somewhat prefer `byte_len` and `byte_offset` to `byte_length`. I agree that `byte_off` is confusing.

Alternatively, `byte_length` could be omitted completely since it's pretty easy to implement.

```
let byte_length = std::mem::size_of::<u32>() * arr.len(&mut cx);
```

The problem with this code is it's really easy for the generic in `size_of` to fall out of sync with the type of `T` in `arr`.

dherman replied:
I could go with `byte_len` and `byte_offset`. But then we just had a discussion and really liked the cleanness of:

```rust
offset()
len()
size()
```

kjvalencik commented on text/0000-typed-array-conveniences.md:
```diff
@@ -0,0 +1,312 @@
+
+## Unfortunate docs placement of `region()`
+
+In order to make the lifetimes work out, `region()` is a method of `Handle<JsArrayBuffer>` instead of `JsArrayBuffer`. This has the unfortunate effect of placing the method in the API docs for [`neon::handle::Handle`][neon::handle::Handle] instead of [`neon::types::JsArrayBuffer`][neon::types::JsArrayBuffer], which makes them less discoverable. The other docs should make an effort to mention the  ...
```
I don't intuitively understand the issue with lifetimes since in `&'a Handle<'cx, Uint8Array>` an `'a` could be chosen that lives as long as `'cx`.

kjvalencik replied:
This is to allow downcasting *and* returning a region in the same method. Otherwise, you would have a compiler error returning a reference to a dropped value.

kjvalencik commented on text/0000-typed-array-conveniences.md:
```diff
@@ -0,0 +1,312 @@
+
+Instead of `buffer.region(offset, len)`, we could expose builder methods for constructing regions, like `buffer.byte_offset(offset).len(len)`. This would have the benefit of making the semantics of the integers more unambiguous, but doesn't seem worth the extra verbosity.
+
+## Alternative names for `Region`
```
I really like `Region`.

kjvalencik reviewed (commented):
This design looks really great! I'm exited for the flexibility this provides.

I left some of my thoughts.

dherman committed:
Changes based on review discussion:
- len()/offset()/size() instead of len()/byte_offset()/byte_length()
- Handle::region() and Region::to_typed_array() take &self
- to_region() -> region()
- add JsArrayBuffer::region() convenience for docs purposes
- neon::types::binary -> neon::types::buffer
- document the alternative idea of compositional regions and why we didn't choose it

kjvalencik reviewed (approved):
❤️ this design!
